### ISSUE TITLE
need guidance on converting/migrating existing hs3 devices to the new hs4 hsfeature

### ISSUE THREAD

**Issue by mnsandler** [HomeSeer/Plugin-SDK#164]
how do i convert all my existing hs3 devices to hs4 features without losing the dv.ref and other properties?

I can create the new hsdevice root device, but I need to then add my existing hs devices to that new root as Features

**Comment by alexbk66** [comment]
I think you set `AssociatedDevices`

**Comment by mnsandler** [comment]
AssociatedDevices just links the root to the features.

But do i have to recreate the hs3 DeviceClass objects to the new hs4 Abstract Device/Feature classes? kind of like we did with actions and triggers.

**Comment by alexbk66** [comment]
I don't think so, you should be able to use EProperty.DeviceType to set ApiType to EApiType.Feature

**Comment by alexbk66** [comment]
I might create a function i.e.

```
void SetParent(int parentRef, int childRef)
{
   SetType(parentRef, EApiType.Device);
   SetAssociatedDevices(parentRef, ...);
   SetType(childRef, EApiType.Feature);
   SetAssociatedDevices(childRef, ...);
}
```

**Comment by mnsandler** [comment]
@alexbk66  
are you suggesting we have to associate a device with features and then also associated features with a device? seems redundant

**Comment by alexbk66** [comment]
> seems redundant
Is it a surprise for you?  That's what I did in my HS3 library:

```
public void SetParent(DeviceBase parent)
{
    parent.AddAssociatedDevice(this);
    parent.SetRelationship(eRelationship.Parent_Root);
    AddAssociatedDevice(parent);
    SetRelationship(eRelationship.Child);
}
```

**Comment by mnsandler** [comment]
is there a 'save' or 'update' method to make all this work? I can't seem to get rid of my old hs3 relationships 

below shows the hs3 associated devices; not the new hs4 root device i'm trying to set

![image](https://user-images.githubusercontent.com/67801043/108604295-06f7b200-737b-11eb-80ee-bcc72f15a13e.png)

**Comment by mnsandler** [comment]
@spudwebb , @alexbk66  

this is really frustrating....

in hs3, i have an existing parent/child relationship (1 to many)

i'm not having any luck converting my HS3 parent device to an HS4 Feature

i'm updating dv.relationship=feature and typeinfo.apitype = feature (seems redudant)

but when i try to run UpdateFeatureByRef, it says the device is a Device type use UpdateDeviceByRef

**is there sample code somewhere that shows how to convert an hs3 parent to an hs4 feature?**

**Comment by mnsandler** [comment]
ok. making progress. here my current issue...i need to convert an existing hs3 device to an hs4 feature.

i can set the Relationship and apiType properties but I have to use the updatedevicebyref to make the changes permanent to convert it to a Feature.  using updatefeaturebyref causes an exception.

is there anyway to know if the current device is a Device or Feature so I can call updatedevicebyref vs updatefeaturebyref?
